### PR TITLE
issue: 4469581 config: add memory size parsing

### PR DESCRIPTION
**Pull request by tomerdbz**
Add support for parsing memory size suffixes (B, KB, MB, GB) in configuration values, enabling users to specify memory limits in human-readable formats like "2GB" instead of byte values.

## Description
- Add memory size transformer supporting B/KB/MB/GB suffixes with whitespace tolerance and overflow detection
- Introduce unified schema analyzer for comprehensive JSON schema property analysis and classification
- Add centralized json_utils module for reusable JSON operations
- Refactor constraint validation to use exceptions with detailed error messages instead of boolean returns
- Update configuration schema to support oneOf patterns for memory parameters with x-memory-size extension flag
- Improve parameter descriptor with value transformation pipeline and enhanced constraint validation

The new architecture eliminates code duplication in schema processing and provides a single source of truth for property analysis. Memory size parsing maintains backward compatibility with plain numeric values while adding user-friendly suffix support.

Extensive unit tests added covering edge cases, overflow handling, whitespace parsing, and integration with existing configuration loaders (JSON and inline).

##### What
Solves https://redmine.mellanox.com/issues/4469581

##### Why ?
Users currently need to specify memory limits in raw byte values (e.g., 2147483648), which is error-prone and hard to read. 

This change allows human-readable formats like "2GB", "512MB", etc., improving user experience and reducing configuration errors. 

The architectural refactoring eliminates significant code duplication when doing so in schema processing and provides better error reporting.

##### How ?
The implementation introduces several key components:

Memory Size Transformer: A robust parser that handles memory size suffixes (B, KB, MB, GB) with whitespace tolerance, overflow detection, and backward compatibility for plain numeric values.

Unified Schema Analyzer: A new schema_analyzer class that provides comprehensive JSON schema property analysis, eliminating duplicate property classification logic and serving as a single source of truth for schema processing.

Centralized JSON Utilities: A json_utils module that consolidates common JSON operations previously scattered across multiple files, improving code reuse and maintainability.

Enhanced Constraint Validation: Refactored constraint validation to use exceptions with detailed error messages instead of boolean returns, providing better debugging information.

Schema Updates: Updated the configuration schema to support oneOf patterns for memory parameters with an x-memory-size extension flag that triggers the memory size transformation.

Value Transformation Pipeline: Enhanced parameter descriptors with a transformation pipeline that applies string mappings first, then value transformations, ensuring proper processing order.

The architecture maintains full backward compatibility while adding the new functionality. Extensive unit tests cover edge cases including overflow handling, whitespace parsing, invalid input handling, and integration with both JSON and inline configuration loaders.

## Change type
What kind of change does this PR introduce?
- [X] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [X] Tests
- [ ] Other

## Check list
- [ ] Code follows the style de facto guidelines of this project
- [ ] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

### CODE REVIEW

**Comment by tomerdbz** [comment]
waiting for dependent PR
https://github.com/Mellanox/libxlio/pull/376

to merge. then will rebase

File: tests/unit_tests/config/config_registry.cpp [hunk 1]
```diff
@@ -321,4 +306,293 @@ TEST(config, config_registry_default_ctr_inline_corrupted_json_ok_throws)
...
+    parameter_descriptor sendfile_param = desc.get_parameter("core.syscall.sendfile_cache_limit");
+
+    // Check the default value type and content
+    std::experimental::any default_val = sendfile_param.default_value();
+
+    // Debug: Check what type the default value actually is
+    std::string type_name = default_val.type().name();
+    std::cout << "Default value type: " << type_name << std::endl;
+
+    // Try different type casts to see what works
```
[the remark this answers is not in the document]
Comment (tomerdbz) [reply tests/unit_tests/config/config_registry.cpp]: fixed.

File: src/core/config/descriptors/parameter_descriptor.cpp [hunk 2]
```diff
@@ -7,6 +7,120 @@
...
+    }
+
+    // Explicitly reject negative values.
+    if (*str == '-') {
+        return false;
+    }
+
+    char *endptr;
+    errno = 0; // Reset errno before the call.
+    unsigned long long value = strtoull(str, &endptr, 10);
```
Comment (AlexanderGrissik) [inline src/core/config/descriptors/parameter_descriptor.cpp:47]: Will strtoull work for string when no space between number and unit like '64G' ?

File: tests/unit_tests/config/config_descriptor.cpp [hunk 3]
```diff
@@ -138,4 +260,4 @@ TEST_F(config_descriptor_test, get_parameter_suggests_cross_section_if_close_eno
         EXPECT_TRUE(error_msg.find("Unknown key 'core.log.lvel'") != std::string::npos);
         EXPECT_TRUE(error_msg.find("Did you mean 'core.log.level'?") != std::string::npos);
     }
-} 
\ No newline at end of file
+}
```
Comment (AlexanderGrissik) [inline tests/unit_tests/config/config_descriptor.cpp:263]: missing new line

File: src/core/config/json_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            const char *str_val = json_object_get_string(enum_value);
+            if (str_val) {
+                values.emplace_back(str_val);
+            }
+        }
+    }
+    return values;
+}
+
+} // namespace json_utils
```
Comment (AlexanderGrissik) [inline src/core/config/json_utils.cpp:186]: missing new line

File: src/core/config/descriptors/parameter_descriptor.cpp [hunk 5]
```diff
@@ -7,6 +7,120 @@
...
+                if (parsed_size > static_cast<uint64_t>(LLONG_MAX)) {
+                    throw_xlio_exception("Memory value '" + str_val +
+                                         "' exceeds maximum supported size.");
+                }
+                return static_cast<int64_t>(parsed_size);
+            }
+        }
+
+        // Return original value if not a string that can be transformed
+        return val;
```
Comment (AlexanderGrissik) [inline src/core/config/descriptors/parameter_descriptor.cpp:121]: If the value cannot be transformed its probably an error that should be displayed to the user. Would this any value propagate to an error?

File: src/core/config/descriptor_providers/schema_analyzer.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    // Helper methods
+    bool has_memory_size_flag();
+    bool has_constraint_fields();
+    bool has_oneof_field();
+    std::experimental::any extract_oneof_value(json_object *one_of_field, std::type_index type,
+                                               const std::string &key);
+
+    json_object *m_property_obj;
+    std::string m_path;
+};
```
Comment (AlexanderGrissik) [inline src/core/config/descriptor_providers/schema_analyzer.h:126]: new line missing

File: tests/unit_tests/config/config_descriptor.cpp (hunk 3 above)
Comment (tomerdbz) [reply tests/unit_tests/config/config_descriptor.cpp:263]: found a newline. what am i missing?

File: src/core/config/descriptors/parameter_descriptor.cpp [hunk 7]
```diff
@@ -7,6 +7,114 @@
...
+ * @return True on successful parsing, false otherwise.
+ */
+bool parameter_descriptor::parse_memory_size(const char *str, uint64_t &size_in_bytes)
+{
+    if (!str) {
+        return false;
+    }
+
+    size_t len = strlen(str);
+    if (len < 2 || str[len - 1] != 'B') {
```
Comment (AlexanderGrissik) [inline src/core/config/descriptors/parameter_descriptor.cpp]: should we trim trailing spaces?

Comment (AlexanderGrissik) [reply src/core/config/descriptors/parameter_descriptor.cpp]: would we support string such as 5G?

File: src/core/config/descriptors/parameter_descriptor.cpp [hunk 8]
```diff
@@ -7,6 +7,114 @@
...
+    }
+
+    // Parse the unit
+    uint64_t multiplier = 1;
+    if (num_len == len - 1) {
+        // Only 'B' at the end, so bytes
+        multiplier = 1;
+    } else if (num_len == len - 2) {
+        switch (str[num_len]) {
+        case 'K':
```
Comment (AlexanderGrissik) [inline src/core/config/descriptors/parameter_descriptor.cpp:59]: what about small case?

File: src/core/config/descriptors/parameter_descriptor.cpp (hunk 2 above)
Comment (tomerdbz) [reply src/core/config/descriptors/parameter_descriptor.cpp]: since this confused you I used stringstream which is clearer :) also solves copilot comment

File: src/core/config/descriptors/parameter_descriptor.cpp (hunk 7 above)
Comment (tomerdbz) [reply src/core/config/descriptors/parameter_descriptor.cpp]: added support
